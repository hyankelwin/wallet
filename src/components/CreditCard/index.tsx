import React from 'react';
import {
  FlagCard,
  SideProps,
} from '../../common/interfaces/CreditCard.interface';
import chip from '../../assets/images/chip.png';
import { Container, CardFront, CardBack, ContentImage } from './styles';
import { mountReplaceNumber } from '../../utils/mountReplaceNumber';

interface CreditCardProps {
  values: {
    number?: string;
    name?: string;
    expiration_month?: string;
    expiration_year?: string;
    cvv?: number;
  };
  side: SideProps;
  fieldFocused: string;
  imageSelected: string;
  flagCard: FlagCard;
}

const CreditCard: React.FC<CreditCardProps> = ({
  values,
  side,
  fieldFocused,
  imageSelected,
  flagCard,
}) => {
  return (
    <Container>
      <CardFront side={side} fieldFocused={fieldFocused} image={imageSelected}>
        <ContentImage>
          <img src={chip} alt="" />
          {flagCard && <img src={flagCard.img} alt="" />}
        </ContentImage>
        <div className="card-number-box">
          {mountReplaceNumber(values.number || '')}
        </div>
        <div className="flexbox">
          <div className="box">
            <span>Card holder</span>
            <div className="card-holder-name">{values.name || 'NOME'}</div>
          </div>
          <div className="box">
            <span>expires</span>
            <div className="expiration">
              <span className="exp-month">
                {values.expiration_month || 'MM'}/
              </span>
              <span className="exp-year">
                {values.expiration_year?.substring(2, 4) || 'YY'}
              </span>
            </div>
          </div>
        </div>
      </CardFront>
      <CardBack side={side} fieldFocused={fieldFocused} image={imageSelected}>
        <div className="stripe" />
        <div className="box">
          <span>cvv</span>
          <div className="cvv-box">{values.cvv}</div>
          {flagCard && <img src={flagCard.img} alt="" />}
        </div>
      </CardBack>
    </Container>
  );
};

export default CreditCard;
