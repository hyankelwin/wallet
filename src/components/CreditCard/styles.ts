import styled, { css } from 'styled-components';
import { SideProps } from '../../common/interfaces/CreditCard.interface';
import Color from '../../styles/color';

interface CreditCardProps {
  side: SideProps;
  fieldFocused: string;
  image: string;
}

export const Container = styled.div`
  position: relative;
  height: 240px;
  width: 400px;
`;

export const CardFront = styled.div<CreditCardProps>`
  position: absolute;
  height: 100%;
  width: 100%;
  left: 30%;
  background: ${props => `url(${props.image})`};
  border-radius: 5px;
  backface-visibility: hidden;
  box-shadow: 0 15px 25px rgba(0, 0, 0, 0.2);
  padding: 20px;
  transition: transform 0.4s ease-out;

  @media (max-width: 600px) {
    left: 20%;
  }

  @media (max-width: 560px) {
    left: 15%;
  }

  @media (max-width: 530px) {
    left: 10%;
  }

  @media (max-width: 500px) {
    left: 7%;
  }

  @media (max-width: 420px) {
    height: 220px;
    width: 350px;
    left: 8%;
  }

  ${props =>
    props.side.value &&
    css`
      transform: perspective(1000px) rotateY(-180deg);
    `};

  ${props =>
    !props.side.value &&
    css`
      transform: perspective(1000px) rotateY(0deg);
    `};

  ${props =>
    !props.image &&
    css`
      background-color: ${Color.primary.default};
    `};

  .card-number-box {
    font-size: 22px;
    color: #fff;
    margin-top: 30px;
    margin-bottom: 30px;

    ${props =>
      props.fieldFocused === 'number' &&
      css`
        font-weight: 500;
      `};

    @media (max-width: 400px) {
      font-size: 18px;
    }
  }

  .flexbox {
    display: flex;
  }

  .box:nth-child(1) {
    margin-right: auto;
  }

  .box {
    font-size: 15px;
    color: #fff;

    @media (max-width: 400px) {
      font-size: 12px;
    }
  }

  .card-holder-name {
    text-transform: uppercase;
  }
`;

export const CardBack = styled.div<CreditCardProps>`
  position: absolute;
  left: 30%;
  height: 100%;
  width: 100%;
  background: ${props => `url(${props.image})`};
  border-radius: 5px;
  padding: 20px 0;
  text-align: right;
  backface-visibility: hidden;
  box-shadow: 0 15px 25px rgba(0, 0, 0, 0.2);
  transform: perspective(1000px) rotateY(180deg);
  transition: transform 0.4s ease-out;

  @media (max-width: 600px) {
    left: 20%;
  }

  @media (max-width: 560px) {
    left: 15%;
  }

  @media (max-width: 530px) {
    left: 10%;
  }

  @media (max-width: 500px) {
    left: 7%;
  }

  @media (max-width: 420px) {
    height: 220px;
    width: 350px;
    left: 8%;
  }

  ${props =>
    props.side.value &&
    css`
      transform: perspective(1000px) rotateY(0deg);
    `};

  ${props =>
    !props.side.value &&
    css`
      transform: perspective(1000px) rotateY(180deg);
    `};

  ${props =>
    !props.image &&
    css`
      background-color: ${Color.primary.default};
    `};

  .stripe {
    background-color: #000;
    width: 100%;
    margin: 10px 0;
    height: 50px;
  }

  .box {
    padding: 0 20px;

    span {
      color: #fff;
      font-size: 15px;
    }

    .cvv-box {
      height: 50px;
      padding: 10px;
      margin-top: 5px;
      color: #333;
      background-color: #fff;
      border-radius: 5px;
      width: 100%;
    }

    img {
      margin-top: 30px;
      height: 30px;

      @media (max-width: 420px) {
        margin-top: 15px;
      }
    }
  }
`;

export const ContentImage = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding-top: 10px;

  img {
    height: 40px;
  }
`;
