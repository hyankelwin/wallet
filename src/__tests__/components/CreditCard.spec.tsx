import { mount } from 'enzyme';
import CreditCard from '../../components/CreditCard';
import { mockCreditCard } from '../../__mocks__/creditCard.mock';
import { mockFlagCard } from '../../__mocks__/flagCard.mock';

describe('CreditCard component', () => {
  describe('should be able to passed in all tests on index.tsx', () => {
    it('should be able to render CreditCard component with success', () => {
      const props = {
        values: mockCreditCard,
        side: { name: 'front', value: false },
        fieldFocused: 'number',
        imageSelected: '/static/img1.png',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should be able to render CreditCard when values.number of mountReplaceNumber is empty', () => {
      const props = {
        values: { ...mockCreditCard, number: '' },
        side: { name: 'front', value: false },
        fieldFocused: 'number',
        imageSelected: '/static/img1.png',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should be able to render CreditCard when values.name is empty', () => {
      const props = {
        values: { ...mockCreditCard, name: '' },
        side: { name: 'front', value: false },
        fieldFocused: 'number',
        imageSelected: '/static/img1.png',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should be able to render CreditCard when values.expiration_month is empty', () => {
      const props = {
        values: { ...mockCreditCard, expiration_month: '' },
        side: { name: 'front', value: false },
        fieldFocused: 'number',
        imageSelected: '/static/img1.png',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should be able to render CreditCard when values.expiration_year is empty', () => {
      const props = {
        values: { ...mockCreditCard, expiration_year: '' },
        side: { name: 'front', value: false },
        fieldFocused: 'number',
        imageSelected: '/static/img1.png',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });
  });

  describe('should be able to passed in all tests on styles.ts', () => {
    it('should be able to render CreditCard when side.value is true', () => {
      const props = {
        values: mockCreditCard,
        side: { name: 'back', value: true },
        fieldFocused: 'number',
        imageSelected: '/static/img1.png',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });

    it('should be able to render CreditCard when side.image is empty', () => {
      const props = {
        values: mockCreditCard,
        side: { name: 'back', value: true },
        fieldFocused: 'number',
        imageSelected: '',
        flagCard: mockFlagCard,
      };
      const wrapper = mount(<CreditCard {...props} />);

      expect(wrapper.debug()).toMatchSnapshot();
    });
  });
});
