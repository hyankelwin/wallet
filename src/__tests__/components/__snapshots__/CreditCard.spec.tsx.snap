// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreditCard component should be able to passed in all tests on index.tsx should be able to render CreditCard component with success 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"/static/img1.png\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-gsDKAQ eChlKK\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            4
            7
            4
            1
             
            5
            8
            5
            4
             
            6
            8
            5
            4
             
            0
            1
            2
            5
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                HYAN KELWIN SILVA FREITAS
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-dkPtRN cfEmyU\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;

exports[`CreditCard component should be able to passed in all tests on index.tsx should be able to render CreditCard when values.expiration_month is empty 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"/static/img1.png\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-gsDKAQ eChlKK\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            4
            7
            4
            1
             
            5
            8
            5
            4
             
            6
            8
            5
            4
             
            0
            1
            2
            5
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                HYAN KELWIN SILVA FREITAS
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-dkPtRN cfEmyU\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;

exports[`CreditCard component should be able to passed in all tests on index.tsx should be able to render CreditCard when values.expiration_year is empty 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"/static/img1.png\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-gsDKAQ eChlKK\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            4
            7
            4
            1
             
            5
            8
            5
            4
             
            6
            8
            5
            4
             
            0
            1
            2
            5
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                HYAN KELWIN SILVA FREITAS
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-dkPtRN cfEmyU\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;

exports[`CreditCard component should be able to passed in all tests on index.tsx should be able to render CreditCard when values.name is empty 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"/static/img1.png\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-gsDKAQ eChlKK\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            4
            7
            4
            1
             
            5
            8
            5
            4
             
            6
            8
            5
            4
             
            0
            1
            2
            5
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                NOME
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-dkPtRN cfEmyU\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;

exports[`CreditCard component should be able to passed in all tests on index.tsx should be able to render CreditCard when values.number of mountReplaceNumber is empty 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"/static/img1.png\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-gsDKAQ eChlKK\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            #
            #
            #
            #
             
            #
            #
            #
            #
             
            #
            #
            #
            #
             
            #
            #
            #
            #
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                HYAN KELWIN SILVA FREITAS
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-dkPtRN cfEmyU\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;

exports[`CreditCard component should be able to passed in all tests on styles.ts should be able to render CreditCard when side.image is empty 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"\\">
        <div className=\\"sc-gsDKAQ jdpFZL\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            4
            7
            4
            1
             
            5
            8
            5
            4
             
            6
            8
            5
            4
             
            0
            1
            2
            5
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                HYAN KELWIN SILVA FREITAS
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"\\">
        <div className=\\"sc-dkPtRN ftitgU\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;

exports[`CreditCard component should be able to passed in all tests on styles.ts should be able to render CreditCard when side.value is true 1`] = `
"<CreditCard values={{...}} side={{...}} fieldFocused=\\"number\\" imageSelected=\\"/static/img1.png\\" flagCard={{...}}>
  <styled.div>
    <div className=\\"sc-bdvvtL cjgPzW\\">
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-gsDKAQ iWIpvS\\">
          <styled.div>
            <div className=\\"sc-hKwDye thxOl\\">
              <img src=\\"chip.png\\" alt=\\"\\" />
              <img src=\\"/static/amex.png\\" alt=\\"\\" />
            </div>
          </styled.div>
          <div className=\\"card-number-box\\">
            4
            7
            4
            1
             
            5
            8
            5
            4
             
            6
            8
            5
            4
             
            0
            1
            2
            5
          </div>
          <div className=\\"flexbox\\">
            <div className=\\"box\\">
              <span>
                Card holder
              </span>
              <div className=\\"card-holder-name\\">
                HYAN KELWIN SILVA FREITAS
              </div>
            </div>
            <div className=\\"box\\">
              <span>
                expires
              </span>
              <div className=\\"expiration\\">
                <span className=\\"exp-month\\">
                  MM
                  /
                </span>
                <span className=\\"exp-year\\">
                  YY
                </span>
              </div>
            </div>
          </div>
        </div>
      </styled.div>
      <styled.div side={{...}} fieldFocused=\\"number\\" image=\\"/static/img1.png\\">
        <div className=\\"sc-dkPtRN koWDSN\\">
          <div className=\\"stripe\\" />
          <div className=\\"box\\">
            <span>
              cvv
            </span>
            <div className=\\"cvv-box\\">
              123
            </div>
            <img src=\\"/static/amex.png\\" alt=\\"\\" />
          </div>
        </div>
      </styled.div>
    </div>
  </styled.div>
</CreditCard>"
`;
